nixos/modules/services/printing/cups-pdf.nix

```diff
@@ -0,0 +1,180 @@
+{ config, lib, pkgs, ... }:
+
+let
+
+  # cups calls its backends as user `lp` (which is good!),
+  # but cups-pdf wants to be called as `root`, so it can change ownership of files.
+  # We add a suid wrapper and a wrapper script to trick cups into calling the suid wrapper.
+
+  # wrapper script that redirects calls to the suid wrapper
+  cups-pdf-wrapper = pkgs.writeTextFile {
+    name = "${pkgs.cups-pdf-to-pdf.name}-wrapper.sh";
+    executable = true;
+    destination = "/lib/cups/backend/cups-pdf";
+    checkPhase = ''
+      ${pkgs.stdenv.shellDryRun} "$target"
+      ${lib.getExe pkgs.shellcheck} "$target"
+    '';
+    text = ''
+      #! ${pkgs.runtimeShell}
+      exec "${config.security.wrapperDir}/cups-pdf" "$@"
+    '';
+  };
```

Comment:
Why do we need this wrapper script if we already have a wrapper? Why can't we use the suid wrapper alone?

Reply:
CUPS is looking for its backend binaries in `$CUPS_PATH/lib/cups/backend/`, but the suid wrapper is placed in `/run/wrappers/bin/cups-pdf`.  So something must redirect CUPS from the former to the latter place.  I also tried this with a symlink, i.e.

```
$CUPS_PATH/lib/cups/backend/cups-pdf -> /run/wrappers/bin/cups-pdf
```

but CUPS complains about this having insecure permissions and refuses to execute it.  Putting a suid wrapper directly into `$CUPS_PATH/lib/cups/backend/` is also not possible since that is a store path.  So apart from completely reorganizing the way CUPS finds its backend binaries, using two wrappers is the only method I can come up with.

I added a note about that implementation detail in the commit message of the corresponding commit so whoever wants to find out how I came up with this construction can look it up.

Reply:
> I added a note about that implementation detail in the commit message of the corresponding commit so whoever wants to find out how I came up with this construction can look it up.

Please move/copy this into a code, otherwise this question will come up again.